pallets/drop3/src/lib.rs

```diff
@@ -0,0 +1,508 @@
+			origin: OriginFor<T>,
+			id: T::PoolId,
+		) -> DispatchResultWithPostInfo {
+			let sender = ensure_signed(origin)?;
+			let mut pool = RewardPools::<T>::get(id).ok_or(Error::<T>::NoSuchRewardPool)?;
+			ensure!(
+				Some(sender.clone()) == Self::admin() || sender == pool.owner,
+				Error::<T>::RequireAdminOrRewardPoolOwner
+			);
+			ensure!(pool.approved, Error::<T>::RewardPoolUnapproved);
+
+			pool.started = true;
+			RewardPools::<T>::insert(id, pool);
+			Self::deposit_event(Event::RewardPoolStarted { id });
+			Ok(().into())
+		}
+
+		/// Stop a reward pool, can be called by admin or reward pool owner
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::stop_reward_pool())]
+		pub fn stop_reward_pool(origin: OriginFor<T>, id: T::PoolId) -> DispatchResultWithPostInfo {
+			let sender = ensure_signed(origin)?;
+			let mut pool = RewardPools::<T>::get(id).ok_or(Error::<T>::NoSuchRewardPool)?;
+			ensure!(
+				Some(sender.clone()) == Self::admin() || sender == pool.owner,
+				Error::<T>::RequireAdminOrRewardPoolOwner
+			);
+			ensure!(pool.approved, Error::<T>::RewardPoolUnapproved);
+
+			pool.started = false;
+			RewardPools::<T>::insert(id, pool);
+			Self::deposit_event(Event::RewardPoolStopped { id });
+			Ok(().into())
+		}
+
+		/// Close a reward pool, can be called by admin or reward pool owner
+		///
+		/// Note here `approved` state is not required, which gives the owner a
+		/// chance to close it before the admin evaluates the proposal
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::close_reward_pool())]
+		pub fn close_reward_pool(
+			origin: OriginFor<T>,
+			id: T::PoolId,
+		) -> DispatchResultWithPostInfo {
+			let sender = ensure_signed(origin)?;
+			let pool = RewardPools::<T>::get(id).ok_or(Error::<T>::NoSuchRewardPool)?;
+			ensure!(
+				Some(sender.clone()) == Self::admin() || sender == pool.owner,
+				Error::<T>::RequireAdminOrRewardPoolOwner
+			);
+
+			Self::unreserve_and_close_reward_pool(id)
+		}
+
+		/// Create a RewardPool proposal, can be called by any signed account
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::propose_reward_pool(T::MaximumNameLength::get()))]
+		pub fn propose_reward_pool(
+			origin: OriginFor<T>,
+			name: Vec<u8>,
+			total: T::Balance,
+			start_at: T::BlockNumber,
+			end_at: T::BlockNumber,
+		) -> DispatchResultWithPostInfo {
+			// a few sanity checks
+			let sender = ensure_signed(origin)?;
+			ensure!(total > 0u32.into(), Error::<T>::InvalidTotalBalance);
+			ensure!(
+				<pallet_balances::Pallet<T> as ReservableCurrency<_>>::can_reserve(&sender, total),
+				Error::<T>::InsufficientReservedBalance
+			);
+			ensure!(end_at >= start_at, Error::<T>::InvalidProposedBlock);
+
+			let bounded_name: BoundedVec<u8, T::MaximumNameLength> =
+				name.clone().try_into().expect("reward pool name is too long");
+
+			// reserve the owner's balance
+			let _ = <pallet_balances::Pallet<T> as ReservableCurrency<_>>::reserve(&sender, total)?;
```

Comment:
An error will be raised here if insufficient balance happens, so the line 366 might be then redundant. (use map_err to use InsufficientReservedBalance error type here.)

Reply:
> An error will be raised here if insufficient balance happens, so the line 366 might be then redundant. (use map_err to use InsufficientReservedBalance error type here.)

Indeed, I'll just leave it out.